### PR TITLE
Function_set can have attr except for chainer.Function

### PR DESCRIPTION
**Pull request by kikusu**
This PR is that Function_set can have attribute except the Function while maintaining the current interface.

### CODE REVIEW

**Comment by beam2d** [comment]
Thanks for PR! I'll review it.

File: chainer/function_set.py [hunk 1]
```diff
@@ -26,8 +26,16 @@ def __init__(self, **functions):
...
+        self.functions = functions
+
+    def __getattr__(self, item):
+        if item == "functions":
+            return super(FunctionSet, self).__getattr__(item)
+
+        if item in self.functions:
+            return self.functions[item]
+
+        raise "'%s' object has no attribute '%s'" % (self.__class__.name, item)
```
Comment (beam2d) [inline chainer/function_set.py]: Wrap the exception by AttributeError.

File: chainer/function_set.py [hunk 2]
```diff
@@ -94,7 +102,8 @@ def parameters(self):
         The order of parameters is consistent with :meth:`gradients` property.
 
         """
-        return sum((func.parameters for _, func in self._get_sorted_funcs()), ())
+        return sum(
+            (func.parameters for _, func in self._get_sorted_funcs()), ())
```
Comment (beam2d) [inline chainer/function_set.py]: Is this line logically changed? If not, revert it to maintain the diff clean.

File: chainer/function_set.py [hunk 3]
```diff
@@ -109,7 +118,8 @@ def gradients(self):
         The order of gradients is consistent with :meth:`parameters` property.
 
         """
-        return sum((func.gradients for _, func in self._get_sorted_funcs()), ())
+        return sum(
+            (func.gradients for _, func in self._get_sorted_funcs()), ())
```
Comment (beam2d) [inline chainer/function_set.py]: ditto.

**Comment by beam2d** [comment]
Would you implement `__setattr__` to accept deferred substitution of Functions and nested FunctionSets? I means following usage must be acceptable.

```python
fs = FunctionSet()
fs.l1 = F.Linear(4, 4)
fs.nested = FunctionSet(...)
```

And we should discuss about the behavior for objects other than Functions and FunctionSets that implement `parameters` and `gradients` properties. Current implementation accepts such objects as attributes of FunctionSet, and their parameters and gradients are collected by FunctionSet.parameters and FunctionSet.gradients.

File: chainer/function_set.py (hunk 2 above)
Comment (kikusu) [inline chainer/function_set.py]: Sorry. There was this line more than 79 characters(pep8 e501). So I have put a new line.

File: chainer/function_set.py (hunk 1 above)
Comment (kikusu) [inline chainer/function_set.py]: Oh, i got a mistake on rebase. I will fix this mistake.

**Comment by kikusu** [comment]
Oh, i forgot such pattern.

```python
class FunctionSet(object):
	def __setattr__(self, key, value):
        if all( [ hasattr(value, i) for i in ["parameters", "gradients"] ]):
        	if "functions" in self.__dict__:
	            self.functions[name] = value
	            return
	    self.__dict__[name] = value

fs = FunctionSet()
fs.l1 = F.Linear(4, 4)
fs.nested = FunctionSet(...)
fs.hoge = 0
print fs.___dict__ # {'functions': {'nested': <chainer.FunctionSet object at >, 'l1': <chainer.functions.linear.Linear object at >}, 'hoge': 0}

```

In this pattern when we use inheritance class from FunctionSet, results are different by the initialization timing of super-class.
```python
class A(FunctionSet):
	def __init__(self):
		super(A, self).__init__()
		self.l= F.Liner(4, 4)
		self.n = 0

a = A()
print a.___dict__ # {'functions': {'l': <chainer.functions.linear.Linear object at >}, 'n': 0}

class B(FunctionSet):
	def __init__(self):
		self.l1= F.Liner(4, 4)
                self.n0 = 0
		super(B, self).__init__() 
		self.l2= F.Liner(4, 4)
                self.n1=0
b = B()
print b.___dict__ # {'functions': {'l2': <chainer.functions.linear.Linear object at >}, 'l1': <chainer.functions.linear.Linear object at >, 'n0':0, 'n1':0}

```

**Comment by beam2d** [comment]
Thank you for the update. I think the second behavior you gave is not intuitive. How about initializing the functions at the first __setattr__ by a function-like object? Also, __init__ should accept non-function attributes.

**Comment by kikusu** [comment]
Thank you for the review. The architecture becomes very complicated when I use `__setattr__`, `__getattr__`.  As a result of having reconsidered it, I want to close this PR.

**Pull request closed without merging**